client/my-sites/plugins/controller-logged-out.js

```diff
@@ -36,11 +42,50 @@ const prefetchPopularPlugins = ( queryClient, options ) =>
 const prefetchFeaturedPlugins = ( queryClient ) =>
 	prefetchPluginsData( queryClient, getWPCOMFeaturedPluginsQueryParams() );
 
-const prefetchProductList = ( store ) => {
-	const productsList = getProductsList( store.getState() );
-	if ( Object.values( productsList ).length === 0 ) {
-		return requestProductsList( { type: 'all' } )( store.dispatch );
+const prefetchProductList = ( queryClient, store ) => {
+	const query = { type: 'all' };
+
+	return queryClient
+		.fetchQuery( [ 'products-list', query.type ], () => wpcom.req.get( '/products', query ) )
+		.then( ( productsList ) => {
+			return store.dispatch( receiveProductsList( productsList, query.type ) );
+		} );
+};
+
+const prefetchTimebox = ( prefetchPromises, context, key, timeout ) => {
+	const racingPromises = [ Promise.all( prefetchPromises ) ];
+	const isBot = context.res?.req?.useragent?.isBot;
+
+	if ( ! timeout ) {
+		timeout = isBot ? PREFETCH_TIMEOUT_BOTS : PREFETCH_TIMEOUT;
 	}
+
+	if ( config.isEnabled( 'ssr/prefetch-timebox' ) ) {
+		const timeboxPromise = new Promise( ( _, reject ) =>
+			setTimeout( reject, timeout, new Error( PREFETCH_TIMEOUT_ERROR ) )
+		);
+
+		racingPromises.push( timeboxPromise );
+	}
+
+	return Promise.race( racingPromises ).catch( ( err ) => {
+		if ( isBot ) {
+			context.res.status( 504 );
+		}
+		context.serverSideRender = false;
+
+		if ( config.isEnabled( 'ssr/log-prefetch-errors' ) ) {
+			logToLogstash( {
```

Comment:
Since this code is running on the server, I'm curious about `logToLogstash`. IIRC, we don't really use that method in the server at all, because we have an existing logfile which is added to logstash. Calling it directly introduces additional network overhead since it needs to ping the logstash server directly. 

Instead, in the server context, we can use `req.logger.error` -- this allows us to output the log message to the Calypso logfile without a network request. That logfile is then parsed into logstash anyways giving us the analytics we need. Plus, the `req.logger` methods have a "request UUID" which is included in the final log, letting you see which logs happened in the same request.

However, since this is a "normal" JS file not in the server, it doesn't have access to the request object. This is out of scope for this PR, but I'm wondering if it'd be valuable to do some magic in `calypso/lib/logstash` which accesses that server logger if we're in the server context. @jsnajdr any thoughts?

Reply:
This is a great observation @noahtallen, I didn't realize at all that using `logToLogstash` might be wrong. On the server, writing to the stdout, which is a structured JSON stream parsed by `bunyan`, is the right thing to do.

The logger should be easily accessible in this function with `context.res.req.logger.error( ... )`. The `req.logger` field is set for every request after `app.use( loggerMiddleware() )`, and `context.res` is set for SSR routes in `getEnhancedContext`.

Reply:
Nice! I didn't realize `context` was available here. I'm curious that since this is in `client/my-sites/plugins/controller-logged-out.js`, will it ever be executed on the client? In which case, we'd probably have problems using that

Reply:
I see that `controller-logged-out.js`, and especially the `fetchSomething` handlers, are imported and called only on server, in the `index.node.js` module.

Reply:
Indeed, this code should only be run on the server. There's also [a check](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plugins/controller-logged-out.js#L49) to make sure the middleware is only running on the server.

Thanks for the insight! 
I replaced it with `context.res.req.logger.error` and also removed the `ssr/log-prefetch-errors` flag check since this doesn't look like it's going to pollute the production logs.

I should mention that I borrowed the `logToLogstash` approach from the [render cache miss](https://github.com/Automattic/wp-calypso/blob/trunk/client/server/render/index.js#L92)  log in `render/index.js` file. Should we use `req.logger.error` there as well?

Reply:
> Should we use `req.logger.error` there as well?

Let's do it in another PR after we confirm that `req.logger` really works 🙂 That `logToLogstash` call predates the addition of `bunyan` logging.